src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -163,60 +156,187 @@ private void CheckSourceLocationSeen(AnonymousTypePublicSymbol anonymous)
+                    template.Construct(typeArguments);
+            }
+            else
+            {
+                var typeParameters = GetReferencedTypeParameters(typeDescr);
+                var key = getTemplateKey(typeDescr, typeParameters);
+
+                // Get anonymous delegate template
+                AnonymousDelegateTemplateSymbol? template;
+                if (!this.SynthesizedDelegates.TryGetValue(key, out template))
+                {
+                    template = this.SynthesizedDelegates.GetOrAdd(key, new AnonymousDelegateTemplateSymbol(this, typeDescr, typeParameters));
+                }
+
+                // Adjust template location if the template is owned by this manager
+                if (ReferenceEquals(template.Manager, this))
+                {
+                    template.AdjustLocation(typeDescr.Location);
                 }
+
+                Debug.Assert(typeParameters.Length == template.TypeParameters.Length);
+                return typeParameters.Length == 0 ?
+                    template :
+                    template.Construct(typeParameters);
             }
-            var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(nTypeArguments);
-            for (int i = 0; i < nTypeArguments; i++)
+
+            static bool allValidTypeArguments(AnonymousTypeDescriptor typeDescr)
             {
-                typeArgumentsBuilder.Add(fields[i].TypeWithAnnotations);
+                var fields = typeDescr.Fields;
+                int n = fields.Length;
+                for (int i = 0; i < n - 1; i++)
+                {
+                    if (!isValidTypeArgument(fields[i]))
+                    {
+                        return true;
+                    }
+                }
+                var returnParameter = fields[n - 1];
+                return !returnParameter.Type.IsVoidType() && !isValidTypeArgument(returnParameter);
             }
-            var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
-            var template = SynthesizeDelegate(parameterCount: fields.Length - 1, refKinds, returnsVoid, generation);
 
-            Debug.Assert(typeArguments.Length == template.TypeParameters.Length);
-            return typeArguments.Length == 0 ?
-                template :
-                template.Construct(typeArguments);
+            static bool isValidTypeArgument(AnonymousTypeField field)
+            {
+                return field.Type is { } type &&
+                    !type.IsPointerOrFunctionPointer() &&
+                    !type.IsRestrictedType();
+            }
+
+            static SynthesizedDelegateKey getTemplateKey(AnonymousTypeDescriptor typeDescr, ImmutableArray<TypeParameterSymbol> typeParameters)
+            {
+                if (typeParameters.Length > 0)
+                {
+                    var typeMap = new TypeMap(typeParameters, IndexedTypeParameterSymbol.Take(typeParameters.Length), allowAlpha: true);
```

Comment:
>allowAlpha: true [](http://example.com/codeflow?start=118&length=16)

Why is this important?

Reply:
`ClosureConversion.RewriteLambdaOrLocalFunction()` rewrites nested functions including the types of any locals in the functions. If the local type is an `AnonymousDelegatePublicSymbol`, that will substitute the types of the associated `TypeDescriptor` with `SynthesizedSubstitutedTypeParameterSymbol` from the rewritten nested function. See `SynthesizedDelegateTypes_30()` for an example.

Reply:
Sorry. This doesn't quite explain to me why ```allowAlpha``` must be true. It looks like we are substituting one set of type parameters with a set of indexed type parameters. How is alpha-rename relevant to this?

Reply:
As we discussed offline, `allowAlpha` is only in the `TypeMap` constructor to allow `SubstitutedTypeParameterSymbol`, and rewritten nested functions currently use `SynthesizedSubstitutedTypeParameterSymbol` instances for type parameters. Perhaps the assert or the rewritten method need to be updated.